File: Spectrogeddon/Spectrogeddon/MobileGLDisplay.m
//
//  MobileGLDisplay.m
//  Spectrogeddon
//
//  Created by Tom York on 25/04/2014.
//  Copyright (c) 2014 Random. All rights reserved.
//

#import "MobileGLDisplay.h"
#import "GLRenderer.h"

@interface MobileGLDisplay ()
@property (nonatomic,strong) EAGLContext* context;
@property (nonatomic,strong) GLRenderer* renderer;
@end

@implementation MobileGLDisplay

- (instancetype)init
{
    if((self = [super init]))
    {
        _context = [[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES2];
        _renderer = [[GLRenderer alloc] init];
    }
    return self;
}

- (void)setGlView:(GLKView *)glView
{
    if(_glView != glView)
    {
        _glView = glView;
        _glView.delegate = self;
        _glView.context = self.context;
    }
}

- (void)useDisplaySettings:(DisplaySettings *)displaySettings
{
    [self.renderer useDisplaySettings:displaySettings];
}

- (void)redisplay
{
    self.renderer.renderSize = (RenderSize) { (GLint)self.glView.bounds.size.width, (GLint)self.glView.bounds.size.height };
    [self.glView display];
}

- (void)glkView:(GLKView *)view drawInRect:(CGRect)rect
{
    [self.renderer render];
}

- (void)addMeasurementToDisplayQueue:(TimeSequence*)timeSequence
{
    [self.renderer addMeasurementsForDisplay:@[ timeSequence ]];
}

@end

File: Spectrogeddon/Spectrogeddon/MobileGLDisplay.swift
#if canImport(GLKit)

//
//  MobileGLDisplay.swift
//  Spectrogeddon
//
//  Created by Tom York on 13/05/2019.
//  Copyright © 2019 Random. All rights reserved.
//

// TODO: Use Metal
@MainActor
@available(iOS, deprecated: 12.0, message: "Uses the legacy OpenGL ES renderer.")
class MobileGLDisplay: NSObject {

    weak var glView: GLKView? {
            didSet {
                glView?.delegate = self
                if let context = context {
                    glView?.context = context
                }
            }
        }
    private var context: EAGLContext?
    private var renderer: GLRenderer
    
    // TODO: Init needs to be failable
    override init() {
        self.context = EAGLContext(api: .openGLES2)
        self.renderer = GLRenderer()
        super.init()
    }
    
    func use(_ displaySettings: DisplaySettings) {
        renderer.use(displaySettings)
    }
    
    func redisplay() {
        guard let view = glView else {
            return
        }

        execute {
            renderer.renderSize = RenderSize(
                width: GLint(view.drawableWidth),
                height: GLint(view.drawableHeight)
            )
            view.display()
        }
    }
    
    func addMeasurement(toDisplayQueue timeSequence: TimeSequence) {
        execute {
            renderer.addMeasurements([timeSequence])
        }
    }

    private func execute(_ commands: () -> Void) {
        guard let context = context else {
            return
        }

        EAGLContext.setCurrent(context)
        commands()
    }
}

@available(iOS, deprecated: 12.0, message: "Uses the legacy OpenGL ES renderer.")
extension MobileGLDisplay: @preconcurrency GLKViewDelegate {
    
    func glkView(_ view: GLKView, drawIn rect: CGRect) {
        execute {
            renderer.render()
        }
    }
}

#endif
